[FEATURE]: Weaviate: Store Metadata + Support gRPC

ehsavoie opened the pull request:
When storing embeddings into Weaviate we are loosing the content metadata. The goal of this PR is to store those metadata as fields in weaviate so that we can get them back when we are finding the content.

## Context
When doing RAG with weaviate as vector database for embeddings we are loosing the metadata associated with the textSegment
## Change
I now store and get back the metadata in Weaviate.
Adding also support for gRPC as according to weaviate team it has better performance.

## Checklist
Before submitting this PR, please check the following points:
- [X] I have added unit and integration tests for my change
- [X] All unit and integration tests in the module I have added/changed are green
- [X] All unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules are green
- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml) (only when a new module is added)

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -30,32 +32,40 @@
  * Represents the <a href="https://weaviate.io/">Weaviate</a> vector database.
- * Current implementation assumes the cosine distance metric is used.
- * Does not support storing {@link dev.langchain4j.data.document.Metadata} yet.
+ * Current implementation assumes the cosine distance metric is used. Does not
```
Please remove "Does not support metadata"

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -30,32 +32,40 @@
     /**
      * Creates a new WeaviateEmbeddingStore instance.
      *
-     * @param apiKey           Your Weaviate API key. Not required for local deployment.
```
please revert previous formatting back

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -100,9 +112,12 @@ public String add(Embedding embedding) {
     /**
      * Adds a new embedding with provided ID to the store.
      *
-     * @param id        the ID of the embedding to add in UUID format, since it's Weaviate requirement.
```
please revert previous formatting back

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -126,31 +141,38 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
     }
 
     /**
-     * {@inheritDoc}
```
please revert previous formatting back

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -213,10 +235,33 @@ private List<String> addAll(List<String> ids, List<Embedding> embeddings, List<T
+        } else {
+            props.put(METADATA_TEXT_SEGMENT, "");
+            for (Property property : properties) {
+                metadata.put(property.getName(), "");
```
this logic seems strange, is it actually required?

ehsavoie replied:
No I should be using "null" instead

dliubars commented on langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/LocalWeaviateEmbeddingStoreIT.java:
```diff
@@ -10,6 +12,14 @@
+import org.junit.jupiter.api.Test;
 
 @Testcontainers
 class LocalWeaviateEmbeddingStoreIT extends EmbeddingStoreWithoutMetadataIT {
```
```suggestion
class LocalWeaviateEmbeddingStoreIT extends EmbeddingStoreIT {
```

dliubars commented on langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/LocalWeaviateEmbeddingStoreIT.java:
```diff
@@ -43,4 +54,34 @@ protected EmbeddingModel embeddingModel() {
\ No newline at end of file
+
+    @Test
+    void should_add_embedding_with_segment_and_metadata() {
```
please remove this test, it is already present in `EmbeddingStoreIT`

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -228,6 +273,15 @@ private WeaviateObject buildObject(String id, Embedding embedding, String text)
+        if (item.get(METADATA) != null && item.get(METADATA) instanceof Map) {
+            Map<String, ?> resultingMetadata = (Map<String, ?>) item.get(METADATA);
+            for(Map.Entry<String, ?> entry : resultingMetadata.entrySet()) {
+                if(entry.getValue() != null && !"".equals(entry.getValue())) {
```
how to know if there was no metadata entry or there was, but with an empty string value?

ehsavoie replied:
Yes, i used "null" instead which seems to be the correct value for null

dliubars commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -65,7 +75,8 @@ public WeaviateEmbeddingStore(
             Boolean avoidDups,
-            String consistencyLevel
+            String consistencyLevel,
+            List<Property> properties
```
```suggestion
            Collection<String> metadataKeys
```
To keep it consistent with other integrations (e.g. redis).

Also, is it absolutely necessary? Is there a way to store/retrieve metadata without specifying explicitly each key?

dliubars reviewed (commented):
@ehsavoie thank you so much!

ehsavoie commented:
@dliubars I think I fixed all your comments.

cescoffier commented:
@dliubars can you have a look, please?

ehsavoie commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -200,34 +227,62 @@ private List<String> addAll(List<String> ids, List<Embedding> embeddings, List<T
+                    if (segment.metadata().containsKey(property.getName())) {
+                        metadata.put(property.getName(), segment.metadata().get(property.getName()));
+                    } else {
+                        metadata.put(property.getName(), "null");
```
the "null" value string is used in the json conversion, otherwise the field is not sent and then the object doesn't respect the schema when trying to obtain it.

ehsavoie commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -135,22 +155,29 @@ public List<EmbeddingMatch<TextSegment>> findRelevant(
+            for (Property property : properties) {
+                metadataFields.add(Field.builder().name(property.getName()).build());
+            }
+            fields.add(Field.builder().name(METADATA).fields(metadataFields.toArray(new Field[0])).build());
```
I'm not sure as I'm not really an expert on this but I don't see why not, you would put ["_metadata", "key"] as the path instead of key in the where filter (as far as I understand the examples on weaviate web site)

ehsavoie replied:
Something like: 
WhereArgument whereUuid = WhereArgument.builder()
      .filter(WhereFilter.builder()
        .path(new String[]{"_metadata", "title"})
        .operator(Operator.Equal)
        .valueText("Man on the Moon")
        .build();

ehsavoie commented on langchain4j-core/src/test/java/dev/langchain4j/store/embedding/EmbeddingStoreIT.java:
How may I reproduce your test ?

ehsavoie replied:
Passing now

ehsavoie commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -200,34 +227,62 @@ private List<String> addAll(List<String> ids, List<Embedding> embeddings, List<T
                     : avoidDups && embedded != null ? generateUUIDFrom(embedded.get(i).text()) : randomUUID();
             resIds.add(id);
-            objects.add(buildObject(id, embeddings.get(i), embedded != null ? embedded.get(i).text() : ""));
+            objects.add(buildObject(id, embeddings.get(i), embedded != null ? embedded.get(i) : null));     
```
Yes, it will work just the same as long as you don't want to get metadata that weren't stored as the fields don't exist

ehsavoie commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -56,30 +61,45 @@ public class WeaviateEmbeddingStore implements EmbeddingStore<TextSegment> {
             );
-            this.client = WeaviateAuthClient.apiKey(config, getOrDefault(apiKey, ""));
+            if (useGrpc) {
+                config.setGRPCSecured(false);
```
Well I used the sample code from Weaviate Client test code and i didn't have mean to set up a secured connection for GRPC. 
I've updated to make this configureable

ehsavoie commented on langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/LocalWeaviateEmbeddingStoreIT.java:
```diff
@@ -1,30 +1,60 @@
             .host(weaviate.getHost())
-            .port(weaviate.getFirstMappedPort())
+            .port(weaviate.getMappedPort(8080))
+            .useGrpc(true)
```
I couldn't figure out what is the GRPC configuration for the cloud testing. Thus I only added a local one

ehsavoie commented on langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/LocalGRPCWeaviateEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,74 @@
+            .scheme("http")
+            .host(weaviate.getHost())
+            .port(weaviate.getMappedPort(8080))
+            .useGrpc(true)
```
Yes you are correct, only batch uses gRPC reading the code from weaviate client

ehsavoie replied:
Confirmed by Weaviate support: only batch inserts use gRPC

ehsavoie replied:
No I don't know about that but it would mean renaming if they change that behaviour. I'd rather stick with weaviate own naming as weaviate doesn't specify that in its configuration

ehsavoie replied:
No I have no problem with that :)
Iwas just wondering if the host should be configurable as i noticed that on thir cloud it has a different host name than the HTTP one

ehsavoie commented on langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/LocalGRPCWeaviateEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,74 @@
+    private final EmbeddingStore<TextSegment> embeddingStore = WeaviateEmbeddingStore.builder()
+            .scheme("http")
+            .host(weaviate.getHost())
+            .port(weaviate.getMappedPort(8080))
```
because weddon't do only inserts

ehsavoie commented on langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java:
```diff
@@ -56,30 +61,47 @@ public class WeaviateEmbeddingStore implements EmbeddingStore<TextSegment> {
             Integer port,
+            Boolean useGrpc,
+            Boolean securedGrpc,
+            Integer grpcPort,
```
Given that we still need HTTP with gRPC I'll keep both.

ehsavoie committed:
Storing content metadata into Weaviate and fetching those fields when getting the content back.
Adding support for gRPC in parallel with HTTP for better performance.
Fixing small issue in EmbeddingStoreIT that was preventing the tests to pass.

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

dliubarskyi committed:
Merge branch 'main' into weaviate-metadata

dliubarskyi committed:
Update langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/LocalWeaviateEmbeddingStoreIT.java
